core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -407,7 +407,9 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
         res = res.recover(cause -> {
             reqCtx.logBuilder().responseCause(cause);
             // Recover the failed response with the error handler.
-            return serviceCfg.errorHandler().onServiceException(reqCtx, cause);
+            try (SafeCloseable ignored = reqCtx.push()) {
+                return serviceCfg.errorHandler().onServiceException(reqCtx, cause);
```

Comment:
note: understood this guarantee is (and only can) be done for `onServiceException` and not the other callback methods